pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go

```diff
@@ -541,17 +554,52 @@ func (lbaas *LbaasV2) UpdateLoadBalancer(clusterName string, service *api.Servic
 		return err
 	}
 
-	// Anything left in addrs is a new member that needs to be added to a pool
-	for addr := range addrs {
-		_, err := v2_pools.CreateAssociateMember(lbaas.network, poolID, v2_pools.MemberCreateOpts{
-			Address:      addr,
-			ProtocolPort: int(ports[0].NodePort),
-			SubnetID:     lbaas.opts.SubnetId,
-		}).Extract()
-		if err != nil {
-			return err
+	waitLoadbalancerActiveProvisioningStatus(lbaas.network, loadbalancer.ID)
+	for _, port := range ports {
+		var listenerID, poolID string
+
+		// Find appropriate listener for this port
+		for _, listener := range listenersList {
+			if listener.ProtocolPort == int(port.Port) && listener.Protocol == string(port.Protocol) {
+				listenerID = listener.ID
+				break
+			}
+		}
+		if listenerID == "" {
+			return fmt.Errorf("Loadbalancer %s does not contain required listener for port %d and protocol %s", loadBalancerName, port.Port, port.Protocol)
+		}
+
+		// Find appropriate pool for found listener
+		for _, pool := range poolsList {
+			for _, currentListener := range pool.Listeners {
+				if listenerID == currentListener.ID {
+					poolID = pool.ID
+					break
+				}
+			}
+			if poolID != "" {
+				break
+			}
+		}
+		if poolID == "" {
+			return fmt.Errorf("Loadbalancer %s does not contain required pool for listener %s", loadBalancerName, listenerID)
+		}
+
+		for addr := range addrs {
```

Comment:
I think using an `addrs` structure shared across all pools/listeners like this fails when the pools get out of sync with each other for some reason.  I think it would be better to have a single loop over `ports` much higher up this function, and do the entire membership update separately for each port/listener/pool.

So: maybe move this "add new members" logic up to the end of the "foreach pool" list at L549-550